Issue/54

hazmat345 opened the pull request:
Fixes #54

loganasherjones commented:
This feels like it should extend to the `SystemClient` API.

hazmat345 commented:
Yep, I agree. Thought about this after submitting the PR. I'll add it.

hazmat345 commented:
Going to WIP this. Need to think more about the intersection of the old `SystemClient` options (`blocking` and `timeout`) and the proposed new ones (`wait`, `max_wait`).

Specifically, I wonder if it makes sense to use `timeout` for both old and new methods.

Also, there are some complications that arise from trying to support both polling and waiting. Specifically, we can't currently support the case where `blocking=False` and `wait=True` - we need an async transport like https://github.com/ross/requests-futures.

It may be easier to just drop polling completely if we introduce async (use async for everything and implement the blocking ourselves for that case). That may remove the need for an API change (except for deprecating `max_delay` since it would no longer have any effect).

loganasherjones commented on brewtils/rest/system_client.py:
```diff
@@ -248,13 +256,16 @@ def send_bg_request(self, **kwargs):
-                                            kwargs.get('_raise_on_error', self._raise_on_error))
+                                            request, raise_on_error)
+
+        # At this point the request will be complete so check for error
```
This breaks backwards compatibility.

The biggest problem I see is if someone is using a newer `brewtils` with an older beer-garden, the `blocking` flag suddenly doesn't work. I think the right thing to do is to introduce some backwards compatibility logic here. I think we should be able to look at the status code for more information (this might come with a change on the API side as well). In any case _something_ returned from the API should tell us the difference between:

* The request has been created, you can query the results here.
* The request was created and successfully executed.

We can then pass the `blocking` flag in the params, if the request is not in bucket 2, then we know we're on an older version of the API and we have to block manually.

hazmat345 commented:
I think this addresses the concerns. Rather than using logic based on additional information I'm essentially just reusing the logic we previously used to ensure a 'blocking' request was finished before it's returned.
